### PR TITLE
[tests-only] test webdav previews

### PR DESCRIPTION
**Pull request by individual-it**
## Description
WebDAV API test for testing the previews of different types of files like text, images, pdf, etc. with valid as well as invalid preview size. 

## Related Issue

#37156 

## How Has This Been Tested?
- locally

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [ ] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by phil-davis** [comment]
@individual-it @swoichha 
PR #37173 has been merge to master. That should have fixed this issue. You will need to rebase this PR and then write the test scenarios so that they pass (with current master you should get "nice" `400` responses)

Remember that the scenarios will fail on older ownCloud. So `skipOnOcV10.3` and `skipOnOcV10.4.0`

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      | simple.pdf   | test.pdf    |
+      | simple.odt   | test.odt    |
+      | new-data.zip | test.zip    |
+
+  Scenario Outline: download previews of different image file types
+    Given user "user0" has uploaded file "filesForUpload/<imageName>" to "/testimage.jpg"
+    When user "user0" downloads the preview of "/testimage.jpg" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "200"
+    And the downloaded image should be "32" pixel wide and "32" pixel high
+    Examples:
```
Comment (individual-it) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature]: these examples are not used

Comment (swoichha) [reply tests/acceptance/features/apiWebdavPreviews/previews.feature]: they are used in line number 71

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    And user "user0" has uploaded file "filesForUpload/lorem.txt" to "/parent.txt"
+    And user "user0" has shared file "/parent.txt" with user "user1"
+    When user "user1" downloads the preview of "/parent.txt" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "200"
+    And the downloaded image should be "32" pixel wide and "32" pixel high
+
+  Scenario: download previews of other users files
+    Given user "user1" has been created with default attributes and without skeleton files
+    And user "user0" has uploaded file "filesForUpload/lorem.txt" to "/parent.txt"
+    When user "user1" downloads the preview of "/parent.txt" with width "32" and height "32" using the WebDAV API
```
Comment (individual-it) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:98]: this step only downloads own previews not other users previews. We need an other step. something like
```suggestion
    When user "user1" downloads the preview of "/parent.txt" of user "user0" with width "32" and height "32" using the WebDAV API
```

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\NotFound"
+
+  Scenario: disable previews
+    Given the administrator has updated system config key "enable_previews" with value "false" and type "boolean"
+    And user "user0" has uploaded file "filesForUpload/lorem.txt" to "/parent.txt"
+    When user "user0" downloads the preview of "/parent.txt" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "404"
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\NotFound"
+
+  Scenario: set maximum size of previews
```
Comment (individual-it) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:123]: this scenario should set   `preview_max_x` & `preview_max_y` and see what happens when a bigger preview is requested
see https://doc.owncloud.com/server/admin_manual/configuration/files/previews_configuration.html#maximum-preview-size

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 4]
```diff
@@ -0,0 +1,129 @@
+@api @TestAlsoOnExternalUserBackend
+Feature: previews of files downloaded thought the webdav API
```
Comment (jasson99) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:2]:
```suggestion
Feature: previews of files downloaded through the webdav API
```

File: tests/TestHelpers/WebDavHelper.php [hunk 5]
```diff
@@ -343,6 +343,8 @@ public static function listFolder(
 	 *                     than download it all up-front.
 	 * @param int $timeout
 	 * @param Client|null $client
+	 * @param array $urlParameter
+	 * @param string $user2
```
Comment (dpakach) [inline tests/TestHelpers/WebDavHelper.php:347]: Please write a line what these arguments are.  Not really clear what it is.

Comment (swoichha) [reply tests/TestHelpers/WebDavHelper.php:347]: Okay. I think `$user2` does not need to be documented. If it should be then can you suggest something?

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+      | %2F    |
+
+  Scenario: download previews of files inside sub-folders
+    Given user "user0" has created folder "subfolder"
+    And user "user0" has uploaded file "filesForUpload/lorem.txt" to "/subfolder/parent.txt"
+    When user "user0" downloads the preview of "/subfolder/parent.txt" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "200"
+    And the downloaded image should be "32" pixel wide and "32" pixel high
+
+  Scenario Outline: download previews of different file types
```
Comment (dpakach) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:59]:
```suggestion
  Scenario Outline: download previews of file types that doesn't support preview
```

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    And the value of the item "/d:error/s:message" in the response should be "Unsupported file type"
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\BadRequest"
+
+  Scenario: download previews of not-existing files
+    When user "user0" downloads the preview of "/parent.txt" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "404"
+    And the value of the item "/d:error/s:message" in the response should be "File with name parent.txt could not be located"
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\NotFound"
+
+  Scenario: disable previews
```
Comment (dpakach) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:116]:
```suggestion
  Scenario: Download file previews when it is disabled by the administrator
```

File: tests/TestHelpers/WebDavHelper.php (hunk 5 above)
Comment (dpakach) [reply tests/TestHelpers/WebDavHelper.php]: Please mention that the DAV request is sent to the dav-path of `$user2`'s file

File: tests/TestHelpers/WebDavHelper.php [hunk 8]
```diff
@@ -360,13 +362,24 @@ public static function makeDavRequest(
 		$authType = "basic",
 		$stream = false,
 		$timeout = 0,
-		$client = null
+		$client = null,
+		$urlParameter = [],
+		$user2 = null
```
Comment (individual-it) [inline tests/TestHelpers/WebDavHelper.php:367]:
```suggestion
		$doDavRequestAsUser = null
```

File: tests/TestHelpers/WebDavHelper.php [hunk 9]
```diff
@@ -343,6 +343,8 @@ public static function listFolder(
 	 *                     than download it all up-front.
 	 * @param int $timeout
 	 * @param Client|null $client
+	 * @param array $urlParameter to concatenate with path
+	 * @param string $user2
```
Comment (individual-it) [inline tests/TestHelpers/WebDavHelper.php:347]:
```suggestion
	 * @param string $doDavRequestAsUser run the DAV as this user, if null its same as $user
```

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    And the value of the item "/d:error/s:message" in the response should be "Unsupported file type"
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\BadRequest"
+
+  Scenario: download previews of not-existing files
+    When user "user0" downloads the preview of "/parent.txt" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "404"
+    And the value of the item "/d:error/s:message" in the response should be "File with name parent.txt could not be located"
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\NotFound"
+
+  Scenario: Download file previews when it disabled by the administrator
```
Comment (Talank) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:116]:
```suggestion
  Scenario: Download file previews when it is disabled by the administrator
```

**Review by kiranparajuli589 — COMMENTED** [review]
lgtm :+1: squash your commits!

**Review by Talank — DISMISSED** [review]
LGTM :+1:

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 11]
```diff
@@ -315,7 +317,9 @@ public function makeDavRequest(
 		$type = "files",
 		$davPathVersion = null,
 		$stream = false,
-		$password = null
+		$password = null,
+		$urlParameter = [],
+		$user2 = null
```
Comment (individual-it) [inline tests/acceptance/features/bootstrap/WebDav.php:322]: lets rename that variable like in the helper

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 12]
```diff
@@ -2925,6 +2929,65 @@ public function userHasDeletedEverythingInFolder($user, $folder) {
...
+			$user, "GET", $path, [], null, "files", 2, false, null, $urlParameter
+		);
+	}
+
+	/**
+	 * @When user :user1 downloads the preview of :path of :user2 with width :width and height :height using the WebDAV API
+	 *
+	 * @param $user1
+	 * @param $path
+	 * @param $user2
```
Comment (individual-it) [inline tests/acceptance/features/bootstrap/WebDav.php:2959]: same here, lets have better variable names

**Review by individual-it — COMMENTED** [review]
better variable names would make the code easier to understand

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+  Scenario: Download file previews when it is disabled by the administrator
+    Given the administrator has updated system config key "enable_previews" with value "false" and type "boolean"
+    And user "user0" has uploaded file "filesForUpload/lorem.txt" to "/parent.txt"
+    When user "user0" downloads the preview of "/parent.txt" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "404"
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\NotFound"
+
+  Scenario: set maximum size of previews
+    Given user "user0" has uploaded file "filesForUpload/lorem.txt" to "/parent.txt"
+    And the administrator has updated system config key "preview_max_x" with value "null"
```
Comment (individual-it) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:128]: need to test also valid max values, but can be done in a separate PR

File: tests/TestHelpers/WebDavHelper.php [hunk 14]
```diff
@@ -403,7 +416,6 @@ public static function makeDavRequest(
 	 * @param int $davPathVersionToUse (1|2)
 	 * @param string $type
 	 *
-	 * @throws InvalidArgumentException
```
Comment (phil-davis) [inline tests/TestHelpers/WebDavHelper.php:406]: what happened here?
Near the end of the function there is `throw new InvalidArgumentException(...`

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 15]
```diff
@@ -2925,6 +2929,65 @@ public function userHasDeletedEverythingInFolder($user, $folder) {
...
+	public function downloadPreview($user, $path, $width, $height) {
+		$urlParameter = [
+			'x' => $width,
+			'y' => $height,
+			'forceIcon' => '0',
+			'preview' => '1'
+		];
+		$this->response = $this->makeDavRequest(
+			$user, "GET", $path, [], null, "files", 2, false, null, $urlParameter
+		);
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebDav.php:2951]:
```suggestion
		$this->downloadPreviewOfOtherUser($user, $path, $user, $width, $height);
```
IMO this will work and avoid duplicating the `$urlParameter` variable setting.

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 16]
```diff
@@ -2925,6 +2929,65 @@ public function userHasDeletedEverythingInFolder($user, $folder) {
...
+	 *
+	 * @param $user1
+	 * @param $path
+	 * @param $doDavRequestAsUser
+	 * @param $width
+	 * @param $height
+	 *
+	 * @return ResponseInterface
+	 */
+	public function downloadPreviewOfOtherUser($user1, $path, $doDavRequestAsUser, $width, $height) {
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebDav.php:2965]: And so to keep our rule that we should try not to directly call step methods from other steps, there could be a separate function with a name like `downloadPreview` - and it can be called by this step, and by the step above. And the step above gets renamed to some long name.

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 17]
```diff
@@ -2925,6 +2929,65 @@ public function userHasDeletedEverythingInFolder($user, $folder) {
...
+			'forceIcon' => '0',
+			'preview' => '1'
+		];
+		$this->response = $this->makeDavRequest(
+			$user1, "GET", $path, [], null, "files", 2, false, null, $urlParameter, $doDavRequestAsUser
+		);
+	}
+
+	/**
+	 * @Then the downloaded image should be :width pixel wide and :height pixel high
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebDav.php:2978]:
```suggestion
	 * @Then the downloaded image should be :width pixels wide and :height pixels high
```

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 18]
```diff
@@ -0,0 +1,132 @@
...
+      | %2F    |
+
+  Scenario: download previews of files inside sub-folders
+    Given user "user0" has created folder "subfolder"
+    And user "user0" has uploaded file "filesForUpload/lorem.txt" to "/subfolder/parent.txt"
+    When user "user0" downloads the preview of "/subfolder/parent.txt" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "200"
+    And the downloaded image should be "32" pixel wide and "32" pixel high
+
+  Scenario Outline: download previews of file types that doesn't support preview
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:60]:
```suggestion
  Scenario Outline: download previews of file types that don't support preview
```

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\NotFound"
+
+  Scenario: Download file previews when it is disabled by the administrator
+    Given the administrator has updated system config key "enable_previews" with value "false" and type "boolean"
+    And user "user0" has uploaded file "filesForUpload/lorem.txt" to "/parent.txt"
+    When user "user0" downloads the preview of "/parent.txt" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "404"
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\NotFound"
+
+  Scenario: set maximum size of previews
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:126]:
```suggestion
  Scenario: unset maximum size of previews
```
This scenario is actually testing that the maximum size of previews can be undefined.

Then, as @individual-it says, in another PR we can add test scenarios that set the max size to various values and check what happens when we ask for a bigger-size preview. Or you can add those now, since there are some changes to do anyway.

**Review by phil-davis — COMMENTED** [review]
A few little things, and suggested refactoring to DRY code a bit.

File: tests/TestHelpers/WebDavHelper.php (hunk 14 above)
Comment (swoichha) [reply tests/TestHelpers/WebDavHelper.php:406]: It must have been removed accidentally.

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+
+  Scenario: unset maximum size of previews
+    Given user "user0" has uploaded file "filesForUpload/lorem.txt" to "/parent.txt"
+    And the administrator has updated system config key "preview_max_x" with value "null"
+    And the administrator has updated system config key "preview_max_y" with value "null"
+    When user "user0" downloads the preview of "/parent.txt" with width "32" and height "32" using the WebDAV API
+    Then the HTTP status code should be "404"
+    And the value of the item "/d:error/s:exception" in the response should be "Sabre\DAV\Exception\NotFound"
+
+  Scenario: set maximum size of previews
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:134]: I don't see where this scenario sets the maximum size of previews

Comment (swoichha) [reply tests/acceptance/features/apiWebdavPreviews/previews.feature:134]: It doesn't set the maximum preview size. It is WIP.

File: tests/acceptance/features/apiWebdavPreviews/previews.feature [hunk 21]
```diff
@@ -0,0 +1,141 @@
...
+    Given user "user0" has uploaded file "filesForUpload/lorem.txt" to "/parent.txt"
+    When the administrator updates system config key "preview_max_x" with value "null" using the occ command
+    And the administrator updates system config key "preview_max_y" with value "null" using the occ command
```
Comment (phil-davis) [inline tests/acceptance/features/apiWebdavPreviews/previews.feature:137]: I thought that the test would be something like:
```suggestion
    When the administrator updates system config key "preview_max_x" with value "32" using the occ command
    And the administrator updates system config key "preview_max_y" with value "32" using the occ command
```

And then try to download a preview of size 64x64.

And see what happens.

**Review by phil-davis — APPROVED** [review]
Let's get this merged, and then extra scenarios can be added in a later PR to test setting the maximum size of previews.

**Comment by phil-davis** [comment]
Ooops - I meant to get this squashed, and forgot before merging. Too late now!

**Comment by phil-davis** [comment]
The new `apiWebdavPreviews` did not get added to `.drone.star` CI. So actually these tests are not running in CI. Fixed in PR #37274
